lib/nvmf/rdma.c

```diff
@@ -3404,15 +3451,57 @@ spdk_nvmf_rdma_poll_group_create(struct spdk_nvmf_transport *transport)
 	}
 
 	TAILQ_INSERT_TAIL(&rtransport->poll_groups, rgroup, link);
+
 	if (rtransport->conn_sched.next_admin_pg == NULL) {
 		rtransport->conn_sched.next_admin_pg = rgroup;
 		rtransport->conn_sched.next_io_pg = rgroup;
 	}
+	rtransport->conn_sched.next_io_pgs[rgroup->numa_node] = rgroup;
 
 	pthread_mutex_unlock(&rtransport->lock);
+
 	return &rgroup->group;
 }
 
+static struct spdk_nvmf_transport_poll_group *
+spdk_nvmf_rdma_get_optimal_numa_poll_group(struct spdk_nvmf_qpair *qpair)
+{
+	struct spdk_nvmf_rdma_transport *rtransport;
+	struct spdk_nvmf_rdma_poll_group **pg;
+	struct spdk_nvmf_transport_poll_group *result;
+	struct spdk_nvmf_rdma_qpair *rqpair;
+
+	rtransport = SPDK_CONTAINEROF(qpair->transport, struct spdk_nvmf_rdma_transport, transport);
+	rqpair = SPDK_CONTAINEROF(qpair, struct spdk_nvmf_rdma_qpair, qpair);
+
+	pthread_mutex_lock(&rtransport->lock);
+
+	if (TAILQ_EMPTY(&rtransport->poll_groups)) {
+		pthread_mutex_unlock(&rtransport->lock);
+		return NULL;
+	}
+
+	if (qpair->qid == 0) {
+		pg = &rtransport->conn_sched.next_admin_pg;
+	} else {
+		pg = &rtransport->conn_sched.next_io_pgs[rqpair->device->numa_node];
+	}
+	assert(*pg != NULL);
```

Comment:
The user can specify reactors mask which will use only 1 numa node (e.g. 0xf - 4 cores on numa 0). But the new connection can be triggered on device attached to numa 1. In this case conn_sched.next_io_pgs[1] will be equal NULL and you'll hit this assert. I think that we should fall back to existing scheduling policy if we don't find a poll group for specific numa node